src/neighbors.cc

```diff
@@ -569,167 +574,88 @@ neighbors_init(void)
     first_ping = Config.peers;
 }
 
-int
-neighborsUdpPing(HttpRequest * request,
-                 StoreEntry * entry,
-                 IRCB * callback,
-                 void *callback_data,
-                 int *exprep,
-                 int *timeout)
+bool
+neighborUdpPing(CachePeer *p,
+                int reqnum,
+                HttpRequest * request,
+                StoreEntry * entry,
+                IRCB * callback,
+                void *callback_data)
 {
-    const char *url = entry->url();
-    MemObject *mem = entry->mem_obj;
-    CachePeer *p = NULL;
-    int i;
-    int reqnum = 0;
     int flags;
     icp_common_t *query;
-    int queries_sent = 0;
-    int peers_pinged = 0;
-    int parent_timeout = 0, parent_exprep = 0;
-    int sibling_timeout = 0, sibling_exprep = 0;
-    int mcast_timeout = 0, mcast_exprep = 0;
-
-    if (Config.peers == NULL)
-        return 0;
-
+    const char *url = entry->url();
+    MemObject *mem = entry->mem_obj;
     assert(entry->swap_status == SWAPOUT_NONE);
+    assert(mem->ping_reply_callback == callback);
+    assert(mem->ircb_data == callback_data);
 
-    mem->start_ping = current_time;
-
-    mem->ping_reply_callback = callback;
-
-    mem->ircb_data = callback_data;
-
-    reqnum = icpSetCacheKey((const cache_key *)entry->key);
-
-    for (i = 0, p = first_ping; i++ < Config.npeers; p = p->next) {
-        if (p == NULL)
-            p = Config.peers;
-
-        debugs(15, 5, "neighborsUdpPing: Peer " << p->host);
-
-        if (!peerWouldBePinged(p, request))
-            continue;       /* next CachePeer */
-
-        ++peers_pinged;
-
-        debugs(15, 4, "neighborsUdpPing: pinging peer " << p->host << " for '" << url << "'");
-
-        debugs(15, 3, "neighborsUdpPing: key = '" << entry->getMD5Text() << "'");
-
-        debugs(15, 3, "neighborsUdpPing: reqnum = " << reqnum);
+    debugs(15, 5, "neighborsUdpPing: Peer " << p->host);
+    debugs(15, 4, "neighborsUdpPing: pinging peer " << p->host << " for '" << url << "'");
+    debugs(15, 3, "neighborsUdpPing: key = '" << entry->getMD5Text() << "'");
+    debugs(15, 3, "neighborsUdpPing: reqnum = " << reqnum);
 
 #if USE_HTCP
-        if (p->options.htcp && !p->options.htcp_only_clr) {
-            if (Config.Port.htcp <= 0) {
-                debugs(15, DBG_CRITICAL, "HTCP is disabled! Cannot send HTCP request to peer.");
-                continue;
-            }
+    if (p->options.htcp && !p->options.htcp_only_clr) {
```

Comment:
I suggest preserving wrong formatting for now, to expose the true code differences.